lib/pbench/server/database/models/user.py

```diff
@@ -90,36 +98,39 @@ def evaluate_role(self, key, value):
     def evaluate_password(self, key, value):
         return generate_password_hash(value)
 
-    # validate the email field
     @validates("email")
     def evaluate_email(self, key, value):
         valid = validate_email(value)
-        email = valid.email
+        return valid.email
 
-        return email
+    def add_token(self, token: ActiveToken):
+        """Add the given token to active tokens list for this user."""
+        try:
+            self.auth_tokens.append(token)
+            Database.db_session.add(token)
+            Database.db_session.commit()
+        except Exception:
+            Database.db_session.rollback()
+            raise
 
     def update(self, **kwargs):
-        """
-        Update the current user object with given keyword arguments
-        """
+        """Update the current user object with given keyword arguments."""
         try:
             for key, value in kwargs.items():
-                if key == "auth_tokens":
-                    # Insert the auth token
-                    self.auth_tokens.append(value)
-                    Database.db_session.add(value)
-                else:
-                    setattr(self, key, value)
+                setattr(self, key, value)
             Database.db_session.commit()
```

Comment:
The current code makes me more nervous than the new code. If we're really *updating* an existing `AuthToken`, then the parent-child relationship is already established, so weren't we re-appending it to the parent relationship list? I'm not entirely sure that *matters* to SQLAlchemy as this list is really just "select * from child where parent = id" and a row can't appear twice in that set. I don't know what the Python `list` will look like in the parent proxy after this block of code, though I suspect that the `commit` will resolve it all ... (but I'm not taking cash bets on that).

And, in any case, while `AuthToken` may or may not survive as a mechanism for "API tokens", the `User` table is definitely going away "soon[er][ish]" and I'm somewhat less inclined to stress over it.

Reply:
> weren't we re-appending it to the parent relationship list?

What I _think_ we were doing (and, this SQLAlchemy stuff makes me question my understanding of everything...) is always adding to and never replacing the token list on the user.  With Peter's changes, we now have a separate interface to the token list (which is probably a good thing), but without the special case here it seems like we could replace the contents of the list with a single token.  That would be awkward, not only from the perspective of losing the other tokens, but also because if we (elsewhere) tried to append to the list we would get an error...and we wouldn't hit this case but in the weird circumstance of doing an add after doing an update.

But, if you think this will work and if we really don't have to worry about it (for very long), I guess I'll just avert my eyes.